File: .sdk-repos/version-8.9/orchestration-cluster-api-js/examples/readme.ts
// Compilable usage examples that appear in README.md.
// These snippets are synced into README.md by scripts/sync-readme-snippets.ts.
// They are type-checked during build (via tsc --noEmit) to guard against API drift.

import path from 'node:path';
import { fileURLToPath } from 'node:url';
import createCamundaClient, {
  createCamundaResultClient,
  isOk,
  isSdkError,
  type JobActionReceipt,
  type JobKey,
  type JobResult,
  type ProcessDefinitionId,
  ProcessDefinitionKey,
  type ProcessInstanceKey,
} from '@camunda8/orchestration-cluster-api';
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Quick Start
// ---------------------------------------------------------------------------

async function _readmeQuickStart() {
  //#region ReadmeQuickStart
  // Zero‑config construction: reads CAMUNDA_* from process.env. If no configuration is present, defaults to Camunda 8 Run on localhost.
  const camunda = createCamundaClient();

  const topology = await camunda.getTopology();
  console.log('Brokers:', topology.brokers?.length ?? 0);
  //#endregion ReadmeQuickStart
}

// ---------------------------------------------------------------------------
// Programmatic Overrides
// ---------------------------------------------------------------------------

async function _readmeOverrides() {
  //#region ReadmeOverrides
  const camunda = createCamundaClient({
    config: {
      CAMUNDA_REST_ADDRESS: 'https://cluster.example',
      CAMUNDA_AUTH_STRATEGY: 'BASIC',
      CAMUNDA_BASIC_AUTH_USERNAME: 'alice',
      CAMUNDA_BASIC_AUTH_PASSWORD: 'secret',
    },
  });
  //#endregion ReadmeOverrides
  void camunda;
}

// ---------------------------------------------------------------------------
// Custom Fetch
// ---------------------------------------------------------------------------

async function _readmeCustomFetch() {
  //#region ReadmeCustomFetch
  const camunda = createCamundaClient({
    fetch: (input, init) => {
      // inspect / modify request here
      return fetch(input, init);
    },
  });
  //#endregion ReadmeCustomFetch
  void camunda;
}

// ---------------------------------------------------------------------------
// Disable Retry
// ---------------------------------------------------------------------------

async function _readmeDisableRetry(jobKey: JobKey) {
  const camunda = createCamundaClient();
  //#region ReadmeDisableRetry
  // This call will not retry on transient errors
  await camunda.completeJob({ jobKey }, { retry: false });
  //#endregion ReadmeDisableRetry
}

// ---------------------------------------------------------------------------
// Override Retry Settings
// ---------------------------------------------------------------------------

async function _readmeRetryOverride(processDefinitionId: ProcessDefinitionId) {
  const camunda = createCamundaClient();
  //#region ReadmeRetryOverride
  // More aggressive retry for this operation only
  await camunda.createProcessInstance(
    { processDefinitionId },
    { retry: { maxAttempts: 8, maxDelayMs: 5000 } }
  );

  // Minimal retry: single retry with short backoff
  await camunda.getTopology({ retry: { maxAttempts: 2, baseDelayMs: 50 } });
  //#endregion ReadmeRetryOverride
}

// ---------------------------------------------------------------------------
// Job Worker Minimal
// ---------------------------------------------------------------------------

async function _readmeJobWorkerMinimal() {
  //#region ReadmeJobWorkerMinimal
  const client = createCamundaClient();

  // Define schemas (optional)
  const Input = z.object({ orderId: z.string() });
  const Output = z.object({ processed: z.boolean() });

  const worker = client.createJobWorker({
    jobType: 'process-order',
    maxParallelJobs: 10,
    jobTimeoutMs: 15_000, // long‑poll timeout (server side requestTimeout)
    pollIntervalMs: 100, // delay between polls when no jobs / at capacity
    // Optional: only fetch specific variables during activation
    fetchVariables: ['orderId'],
    inputSchema: Input, // validates incoming variables if validateSchemas true
    outputSchema: Output, // validates variables passed to complete(...)
    validateSchemas: true, // set false for max throughput (skip Zod)
    autoStart: true, // default true; start polling immediately
    startupJitterMaxSeconds: 5, // random delay up to 5s before first poll (default 0)
    jobHandler: (job) => {
      // Access typed variables
      const vars = job.variables; // inferred from Input schema
      console.log(`Processing order: ${vars.orderId}`);
      // Do work...
      return job.complete({ processed: true });
    },
  });

  // Later, on shutdown:
  process.on('SIGINT', () => {
    worker.stop();
  });
  //#endregion ReadmeJobWorkerMinimal
}

// ---------------------------------------------------------------------------
// Job Worker TypeScript Inference
// ---------------------------------------------------------------------------

async function _readmeJobWorkerInference() {
  const client = createCamundaClient();
  //#region ReadmeJobWorkerInference
  const Input = z.object({ orderId: z.string(), amount: z.number() });
  client.createJobWorker({
    jobType: 'process-order',
    maxParallelJobs: 5,
    jobTimeoutMs: 30_000,
    inputSchema: Input,
    // Only allows 'orderId' | 'amount' here at compile-time
    fetchVariables: ['orderId', 'amount'],
    jobHandler: async (job) => job.complete(),
  });
  //#endregion ReadmeJobWorkerInference
}

// ---------------------------------------------------------------------------
// Job Worker Startup Jitter
// ---------------------------------------------------------------------------

async function _readmeJobWorkerJitter() {
  const client = createCamundaClient();
  //#region ReadmeJobWorkerJitter
  client.createJobWorker({
    jobType: 'process-order',
    maxParallelJobs: 10,
    jobTimeoutMs: 30_000,
    startupJitterMaxSeconds: 5, // each instance delays 0–5s before first poll
    jobHandler: async (job) => job.complete(),
  });
  //#endregion ReadmeJobWorkerJitter
}

// ---------------------------------------------------------------------------
// Job Worker Graceful Shutdown
// ---------------------------------------------------------------------------

async function _readmeJobWorkerGraceful() {
  const client = createCamundaClient();
  const worker = client.createJobWorker({
    jobType: 'process-order',
    maxParallelJobs: 5,
    jobTimeoutMs: 30_000,
    jobHandler: async (job) => job.complete(),
  });

  //#region ReadmeJobWorkerGraceful
  // Attempt graceful drain for up to 8 seconds
  const { remainingJobs, timedOut } = await worker.stopGracefully({ waitUpToMs: 8000 });
  if (timedOut) {
    console.warn('Graceful stop timed out; remaining jobs:', remainingJobs);
  }
  //#endregion ReadmeJobWorkerGraceful
}

// ---------------------------------------------------------------------------
// Job Action Receipt
// ---------------------------------------------------------------------------

async function _readmeReceipt() {
  const client = createCamundaClient();
  const worker = client.createJobWorker({
    jobType: 'process-order',
    maxParallelJobs: 5,
    jobTimeoutMs: 30_000,
    jobHandler: async (job) => {
      //#region ReadmeReceipt
      const receipt: JobActionReceipt = await job.complete({ processed: true });
      //#endregion ReadmeReceipt
      return receipt;
    },
  });
  void worker;
}

// ---------------------------------------------------------------------------
// Job Corrections (User Task Listeners)
// ---------------------------------------------------------------------------

async function _readmeJobCorrections() {
  const client = createCamundaClient();
  //#region ReadmeJobCorrections
  const worker = client.createJobWorker({
    jobType: 'io.camunda:userTaskListener',
    jobTimeoutMs: 30_000,
    maxParallelJobs: 5,
    jobHandler: async (job) => {
      const result: JobResult = {
        type: 'userTask',
        corrections: {
          assignee: 'corrected-user',
          priority: 80,
        },
      };
      return job.complete({}, result);
    },
  });
  //#endregion ReadmeJobCorrections
  void worker;
}

async function _readmeJobCorrectionsDenial() {
  const client = createCamundaClient();
  const worker = client.createJobWorker({
    jobType: 'io.camunda:userTaskListener',
    jobTimeoutMs: 30_000,
    maxParallelJobs: 5,
    jobHandler: async (job) => {
      //#region ReadmeJobCorrectionsDenial
      return job.complete(
        {},
        {
          type: 'userTask',
          denied: true,
          deniedReason: 'Insufficient documentation',
        }
      );
      //#endregion ReadmeJobCorrectionsDenial
    },
  });
  void worker;
}

// ---------------------------------------------------------------------------
// Threaded Job Handler Module
// ---------------------------------------------------------------------------

// NOTE: The ReadmeThreadedHandler snippet in the README is pseudo-code
// (references heavyComputation, uses export default). It cannot be
// type-checked in this file, so it is not auto-synced.

// ---------------------------------------------------------------------------
// Threaded Job Worker Minimal
// ---------------------------------------------------------------------------

async function _readmeThreadedWorker() {
  //#region ReadmeThreadedWorker
  const client = createCamundaClient();

  const worker = client.createThreadedJobWorker({
    jobType: 'cpu-heavy-task',
    handlerModule: path.join(path.dirname(fileURLToPath(import.meta.url)), 'my-handler.js'),
    maxParallelJobs: 32,
    jobTimeoutMs: 30_000,
  });
  //#endregion ReadmeThreadedWorker
  void worker;
}

// ---------------------------------------------------------------------------
// Branded Keys
// ---------------------------------------------------------------------------

function _readmeBrandedKeys() {
  //#region ReadmeBrandedKeys
  const defKey = ProcessDefinitionKey.assumeExists('2251799813686749');
  // @ts-expect-error – cannot assign def key to instance key
  const bad: ProcessInstanceKey = defKey;
  //#endregion ReadmeBrandedKeys
  void bad;
}

// ---------------------------------------------------------------------------
// Cancelable Operations
// ---------------------------------------------------------------------------

async function _readmeCancelable(defKey: ProcessDefinitionKey) {
  const camunda = createCamundaClient();
  //#region ReadmeCancelable
  const p = camunda.searchProcessInstances(
    { filter: { processDefinitionKey: defKey } },
    { consistency: { waitUpToMs: 0 } }
  );
  setTimeout(() => p.cancel(), 100); // best‑effort cancel
  try {
    await p; // resolves if not cancelled
  } catch (e) {
    if (isSdkError(e) && e.name === 'CancelSdkError') {
      console.log('Operation cancelled');
    } else throw e;
  }
  //#endregion ReadmeCancelable
}

// ---------------------------------------------------------------------------
// Eventual Consistency
// ---------------------------------------------------------------------------

async function _readmeEventualConsistency() {
  const camunda = createCamundaClient();
  //#region ReadmeEventualConsistency
  const jobs = await camunda.searchJobs(
    {
      filter: { type: 'payment' },
    },
    {
      consistency: {
        waitUpToMs: 5000,
        pollIntervalMs: 200,
        trace: true,
        predicate: (r) => Array.isArray(r.items) && r.items.some((j) => j.state === 'CREATED'),
      },
    }
  );
  //#endregion ReadmeEventualConsistency
  void jobs;
}

// ---------------------------------------------------------------------------
// Logging
// ---------------------------------------------------------------------------

async function _readmeLogging() {
  //#region ReadmeLogging
  const client = createCamundaClient({
    log: {
      level: 'info',
      transport: (evt) => {
        // evt: { level, scope, ts, args, code?, data? }
        console.log(JSON.stringify(evt));
      },
    },
  });

  const log = client.logger('worker');
  log.debug(() => ['expensive detail only if enabled', { meta: 1 }]);
  log.code('info', 'WORK_START', 'Starting work loop', { pid: process.pid });
  //#endregion ReadmeLogging
}

// ---------------------------------------------------------------------------
// Typed Error Handling
// ---------------------------------------------------------------------------

async function _readmeErrorHandling() {
  //#region ReadmeErrorHandling
  const client = createCamundaClient();

  try {
    await client.getTopology();
  } catch (e) {
    if (isSdkError(e)) {
      switch (e.name) {
        case 'HttpSdkError':
          console.error('HTTP failure', e.status, e.operationId);
          break;
        case 'ValidationSdkError':
          console.error('Validation issue on', e.operationId, e.side, e.issues);
          break;
        case 'AuthSdkError':
          console.error('Auth problem', e.message, e.status);
          break;
        case 'CancelSdkError':
          console.error('Operation cancelled', e.operationId);
          break;
        case 'NetworkSdkError':
          console.error('Network layer error', e.message);
          break;
      }
      return;
    }
    // Non-SDK (programmer) error; rethrow or wrap
    throw e;
  }
  //#endregion ReadmeErrorHandling
}

// ---------------------------------------------------------------------------
// Result Client
// ---------------------------------------------------------------------------

async function _readmeResultClient() {
  const file = new File(['<xml/>'], 'process.bpmn', { type: 'application/xml' });
  //#region ReadmeResultClient
  const camundaR = createCamundaResultClient();
  const res = await camundaR.createDeployment({ resources: [file] });
  if (isOk(res)) {
    console.log('Deployment key', res.value.deploymentKey);
  } else {
    console.error('Deployment failed', res.error);
  }
  //#endregion ReadmeResultClient
}

// ---------------------------------------------------------------------------
// Browser Deployment
// ---------------------------------------------------------------------------

async function _readmeDeployBrowser() {
  const camunda = createCamundaClient();
  //#region ReadmeDeployBrowser
  const bpmnXml = `<definitions id="process" xmlns="http://www.omg.org/spec/BPMN/20100524/MODEL">...</definitions>`;
  const file = new File([bpmnXml], 'order-process.bpmn', { type: 'application/xml' });
  const result = await camunda.createDeployment({ resources: [file] });
  console.log(result.deployments.length);
  //#endregion ReadmeDeployBrowser
}

// ---------------------------------------------------------------------------
// Node Deployment
// ---------------------------------------------------------------------------

async function _readmeDeployNode() {
  const camunda = createCamundaClient();
  //#region ReadmeDeployNode
  const result = await camunda.deployResourcesFromFiles([
    './bpmn/order-process.bpmn',
    './dmn/discount.dmn',
    './forms/order.form',
  ]);

  console.log(result.processes.map((p) => p.processDefinitionId));
  console.log(result.decisions.length);
  //#endregion ReadmeDeployNode
}

// ---------------------------------------------------------------------------
// Testing Patterns
// ---------------------------------------------------------------------------

async function _readmeTestingClient() {
  //#region ReadmeTestingClient
  const client = createCamundaClient({
    config: { CAMUNDA_REST_ADDRESS: 'http://localhost:8080', CAMUNDA_AUTH_STRATEGY: 'NONE' },
  });
  //#endregion ReadmeTestingClient
  void client;
}

async function _readmeTestingMock() {
  //#region ReadmeTestingMock
  const client = createCamundaClient({
    fetch: async (_input, _init) => new Response(JSON.stringify({ ok: true }), { status: 200 }),
  });
  //#endregion ReadmeTestingMock
  void client;
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
// Heritable Worker Defaults
// ---------------------------------------------------------------------------

async function _readmeWorkerDefaultsEnv() {
  const client = createCamundaClient();

  //#region ReadmeWorkerDefaultsEnv
  // Workers inherit timeout, concurrency, and name from environment
  const w1 = client.createJobWorker({
    jobType: 'validate-order',
    jobHandler: async (job) => job.complete(),
  });

  const w2 = client.createJobWorker({
    jobType: 'ship-order',
    jobHandler: async (job) => job.complete(),
  });

  // Per-worker override: this worker uses 32 concurrent jobs instead of the global 8
  const w3 = client.createJobWorker({
    jobType: 'bulk-import',
    maxParallelJobs: 32,
    jobHandler: async (job) => job.complete(),
  });
  //#endregion ReadmeWorkerDefaultsEnv

  void w1;
  void w2;
  void w3;
}

async function _readmeWorkerDefaultsClient() {
  //#region ReadmeWorkerDefaultsClient
  const client = createCamundaClient({
    config: {
      CAMUNDA_WORKER_TIMEOUT: 30000,
      CAMUNDA_WORKER_MAX_CONCURRENT_JOBS: 8,
    },
  });
  //#endregion ReadmeWorkerDefaultsClient
  void client;
}

// ---------------------------------------------------------------------------
// Job Completion Patterns
// ---------------------------------------------------------------------------

async function _readmeJobCompletionPatterns() {
  const client = createCamundaClient();
  const worker = client.createJobWorker({
    jobType: 'example',
    maxParallelJobs: 1,
    jobTimeoutMs: 30_000,
    jobHandler: async (job) => {
      //#region ReadmeJobCompletionPatterns
      // GOOD: explicit completion
      return job.complete({ variables: { processed: true } });

      // GOOD: No-arg completion example, sentinel stored for ultimate return
      const ack = await job.complete();
      // ...
      return ack;

      // GOOD: explicit ignore
      const ack2 = await job.ignore();
      //#endregion ReadmeJobCompletionPatterns
      return ack2;
    },
  });
  void worker;
}

// ---------------------------------------------------------------------------
// Backpressure State Shape
// ---------------------------------------------------------------------------

function _readmeBackpressureState() {
  const client = createCamundaClient();
  //#region ReadmeBackpressureState
  const state = client.getBackpressureState();
  // state.severity: 'healthy' | 'soft' | 'severe'
  // state.consecutive: number — consecutive backpressure signals observed
  // state.permitsMax: number | null — current concurrency cap (null => unlimited/not engaged)
  // state.permitsCurrent: number — currently acquired permits
  // state.waiters: number — queued operations waiting for a permit
  //#endregion ReadmeBackpressureState
  void state;
}

// ---------------------------------------------------------------------------
// Threaded Worker Lifecycle
// ---------------------------------------------------------------------------

async function _readmeThreadedLifecycle() {
  const client = createCamundaClient();
  //#region ReadmeThreadedLifecycle
  // Returned by getWorkers()
  const allWorkers = client.getWorkers();

  // Stopped by stopAllWorkers()
  client.stopAllWorkers();
  //#endregion ReadmeThreadedLifecycle
  void allWorkers;
}

async function _readmeThreadedGraceful() {
  const client = createCamundaClient();
  const worker = client.createJobWorker({
    jobType: 'example',
    maxParallelJobs: 1,
    jobTimeoutMs: 30_000,
    jobHandler: async (job) => job.complete(),
  });
  //#region ReadmeThreadedGraceful
  // Graceful shutdown (waits for in-flight jobs to finish)
  const { timedOut, remainingJobs } = await worker.stopGracefully({ waitUpToMs: 10_000 });
  //#endregion ReadmeThreadedGraceful
  void timedOut;
  void remainingJobs;
}

function _readmePoolStats() {
  const client = createCamundaClient();
  const worker = client.createThreadedJobWorker({
    jobType: 'example',
    handlerModule: './handler.js',
    maxParallelJobs: 1,
    jobTimeoutMs: 30_000,
  });
  //#region ReadmePoolStats
  worker.poolSize; // number of threads
  worker.busyThreads; // threads currently processing a job
  worker.activeJobs; // total jobs dispatched but not yet completed
  //#endregion ReadmePoolStats
}

// Suppress "declared but never read"
void _readmeQuickStart;
void _readmeOverrides;
void _readmeCustomFetch;
void _readmeDisableRetry;
void _readmeRetryOverride;
void _readmeJobWorkerMinimal;
void _readmeJobWorkerInference;
void _readmeJobWorkerJitter;
void _readmeJobWorkerGraceful;
void _readmeReceipt;
void _readmeThreadedWorker;
void _readmeBrandedKeys;
void _readmeCancelable;
void _readmeEventualConsistency;
void _readmeLogging;
void _readmeErrorHandling;
void _readmeResultClient;
void _readmeDeployBrowser;
void _readmeDeployNode;
void _readmeTestingClient;
void _readmeTestingMock;
void _readmeWorkerDefaultsEnv;
void _readmeWorkerDefaultsClient;
void _readmeJobCompletionPatterns;
void _readmeBackpressureState;
void _readmeThreadedLifecycle;
void _readmeThreadedGraceful;
void _readmePoolStats;

File: .sdk-repos/version-8.9/orchestration-cluster-api-python/examples/readme.py
"""README code examples — type-checked during build to ensure accuracy.

These snippets are synced into README.md by scripts/sync-readme-snippets.py.
Edit these examples to update the corresponding README code blocks.
"""

# pyright: reportUnusedVariable=false, reportUnusedFunction=false
# ruff: noqa: F841

from __future__ import annotations


# ---------- Using the SDK ----------


def readme_sync_client() -> None:
    # region ReadmeSyncClient
    # Sync
    from camunda_orchestration_sdk import CamundaClient

    with CamundaClient() as client:
        topology = client.get_topology()
    # endregion ReadmeSyncClient


def readme_async_client() -> None:
    # region ReadmeAsyncClient
    # Async
    import asyncio

    from camunda_orchestration_sdk import CamundaAsyncClient

    async def main() -> None:
        async with CamundaAsyncClient() as client:
            topology = await client.get_topology()

    asyncio.run(main())
    # endregion ReadmeAsyncClient


# ---------- Semantic Types ----------


def readme_semantic_types() -> None:
    # region ReadmeSemanticTypes
    from camunda_orchestration_sdk import CamundaClient, ProcessCreationByKey

    client = CamundaClient()

    # Deploy → the response already carries typed keys
    deployment = client.deploy_resources_from_files(["process.bpmn"])
    process_key = deployment.processes[0].process_definition_key  # ProcessDefinitionKey

    # Pass it directly to another call — no conversion needed
    result = client.create_process_instance(
        data=ProcessCreationByKey(process_definition_key=process_key)
    )

    # The result also carries typed keys
    instance_key = result.process_instance_key  # ProcessInstanceKey
    client.cancel_process_instance(process_instance_key=instance_key)
    # endregion ReadmeSemanticTypes


# ---------- Quick start ----------


def readme_zero_config() -> None:
    # region ReadmeZeroConfig
    from camunda_orchestration_sdk import CamundaAsyncClient, CamundaClient

    # Zero-config construction: reads CAMUNDA_* from the environment
    client = CamundaClient()
    async_client = CamundaAsyncClient()
    # endregion ReadmeZeroConfig


# ---------- .env loading ----------


def readme_env_file_loading() -> None:
    # region ReadmeEnvFileLoading
    from camunda_orchestration_sdk import CamundaClient

    client = CamundaClient(configuration={"CAMUNDA_LOAD_ENVFILE": "true"})
    # endregion ReadmeEnvFileLoading


# ---------- Programmatic configuration ----------


def readme_programmatic_config() -> None:
    # region ReadmeProgrammaticConfig
    from camunda_orchestration_sdk import CamundaClient

    client = CamundaClient(
        configuration={
            "CAMUNDA_REST_ADDRESS": "http://localhost:8080/v2",
            "CAMUNDA_AUTH_STRATEGY": "NONE",
        }
    )
    # endregion ReadmeProgrammaticConfig


# ---------- Basic auth ----------


def readme_basic_auth() -> None:
    # region ReadmeBasicAuth
    from camunda_orchestration_sdk import CamundaClient

    client = CamundaClient(
        configuration={
            "CAMUNDA_REST_ADDRESS": "http://localhost:8080/v2",
            "CAMUNDA_AUTH_STRATEGY": "BASIC",
            "CAMUNDA_BASIC_AUTH_USERNAME": "your-username",
            "CAMUNDA_BASIC_AUTH_PASSWORD": "your-password",
        }
    )
    # endregion ReadmeBasicAuth


# ---------- Deploying Resources ----------


def readme_deploy_resources() -> None:
    # region ReadmeDeployResources
    from camunda_orchestration_sdk import CamundaClient

    with CamundaClient() as client:
        result = client.deploy_resources_from_files(["process.bpmn", "decision.dmn"])

        print(f"Deployment key: {result.deployment_key}")
        for process in result.processes:
            print(f"  Process: {process.process_definition_id} (key: {process.process_definition_key})")
    # endregion ReadmeDeployResources


# ---------- Creating a Process Instance ----------


def readme_create_process_instance() -> None:
    # region ReadmeCreateProcessInstance
    from camunda_orchestration_sdk import CamundaClient, ProcessCreationByKey

    with CamundaClient() as client:
        # Deploy and capture the typed key
        deployment = client.deploy_resources_from_files(["process.bpmn"])
        process_key = deployment.processes[0].process_definition_key

        # Use it directly — the type flows through without conversion
        result = client.create_process_instance(
            data=ProcessCreationByKey(process_definition_key=process_key)
        )
        print(f"Process instance key: {result.process_instance_key}")
    # endregion ReadmeCreateProcessInstance


def readme_create_from_storage() -> None:
    # region ReadmeCreateFromStorage
    from camunda_orchestration_sdk import CamundaClient, ProcessCreationByKey, ProcessDefinitionKey

    with CamundaClient() as client:
        stored_key = "2251799813685249"  # from a DB row or config
        result = client.create_process_instance(
            data=ProcessCreationByKey(process_definition_key=ProcessDefinitionKey(stored_key))
        )
        print(f"Process instance key: {result.process_instance_key}")
    # endregion ReadmeCreateFromStorage


# ---------- Job Workers ----------


def readme_job_worker() -> None:
    # region ReadmeJobWorker
    import asyncio

    from camunda_orchestration_sdk import CamundaAsyncClient, ConnectedJobContext, WorkerConfig

    async def handle_job(job_context: ConnectedJobContext) -> dict[str, object]:
        variables = job_context.variables.to_dict()
        job_context.log.info(f"Processing job {job_context.job_key}: {variables}")
        return {"result": "processed"}

    async def main() -> None:
        async with CamundaAsyncClient() as client:
            config = WorkerConfig(
                job_type="my-service-task",
                job_timeout_milliseconds=30_000,
            )
            client.create_job_worker(config=config, callback=handle_job)

            # Keep workers running until cancelled
            await client.run_workers()

    asyncio.run(main())
    # endregion ReadmeJobWorker


# ---------- Using the Client in a Job Handler ----------


async def readme_async_handler() -> None:
    # region ReadmeAsyncHandler
    from camunda_orchestration_sdk import ConnectedJobContext, MessagePublicationRequest, MessagePublicationRequestVariables

    async def handle_order(job: ConnectedJobContext) -> dict[str, object]:
        variables = job.variables.to_dict()
        order_id = variables["orderId"]

        await job.client.publish_message(
            data=MessagePublicationRequest(
                name="order-processed",
                correlation_key=order_id,
                time_to_live=60000,
                variables=MessagePublicationRequestVariables.from_dict({"orderId": order_id, "status": "completed"}),
            )
        )

        job.log.info(f"Published order-processed message for order {order_id}")
        return {"status": "done"}
    # endregion ReadmeAsyncHandler


def readme_sync_handler() -> None:
    # region ReadmeSyncHandler
    from camunda_orchestration_sdk import MessagePublicationRequest, MessagePublicationRequestVariables, SyncJobContext

    def handle_order(job: SyncJobContext) -> dict[str, object]:
        variables = job.variables.to_dict()
        order_id = variables["orderId"]

        job.client.publish_message(
            data=MessagePublicationRequest(
                name="order-processed",
                correlation_key=order_id,
                time_to_live=60000,
                variables=MessagePublicationRequestVariables.from_dict({"orderId": order_id, "status": "completed"}),
            )
        )

        job.log.info(f"Published order-processed message for order {order_id}")
        return {"status": "done"}
    # endregion ReadmeSyncHandler


# ---------- Failing a Job ----------


async def readme_fail_job() -> None:
    # region ReadmeFailJob
    from camunda_orchestration_sdk import ConnectedJobContext, JobFailure

    async def handle_job(job: ConnectedJobContext) -> dict[str, object]:
        if not job.variables.to_dict().get("required_field"):
            raise JobFailure(
                message="Missing required field",
                retries=2,
                retry_back_off=5000,  # milliseconds
            )
        return {"result": "ok"}
    # endregion ReadmeFailJob


# ---------- BPMN Error ----------


async def readme_bpmn_error() -> None:
    # region ReadmeBpmnError
    from camunda_orchestration_sdk import ConnectedJobContext, JobError

    async def handle_payment(job: ConnectedJobContext) -> dict[str, object]:
        variables = job.variables.to_dict()
        if variables.get("amount", 0) > 10_000:
            raise JobError(error_code="AMOUNT_TOO_HIGH", message="Payment exceeds limit")
        return {"status": "approved"}
    # endregion ReadmeBpmnError


# ---------- Error Handling ----------


def readme_error_handling() -> None:
    # region ReadmeErrorHandling
    from camunda_orchestration_sdk import CamundaClient, ProcessCreationByKey, ProcessDefinitionKey
    from camunda_orchestration_sdk.errors import BadRequestError

    process_definition_key = ProcessDefinitionKey("2251799813685249")

    with CamundaClient() as client:
        try:
            result = client.create_process_instance(
                data=ProcessCreationByKey(process_definition_key=process_definition_key)
            )
        except BadRequestError as e:
            print(f"Bad request ({e.operation_id}): {e}")
    # endregion ReadmeErrorHandling


# ---------- Logging ----------


def readme_stdlib_logger() -> None:
    # region ReadmeStdlibLogger
    import logging

    from camunda_orchestration_sdk import CamundaClient

    my_logger = logging.getLogger("my_app.camunda")
    my_logger.setLevel(logging.DEBUG)

    client = CamundaClient(logger=my_logger)
    # endregion ReadmeStdlibLogger


def readme_custom_logger() -> None:
    # region ReadmeCustomLogger
    from camunda_orchestration_sdk import CamundaClient

    class MyLogger:
        def debug(self, msg: object, *args: object, **kwargs: object) -> None:
            print(f"[DEBUG] {msg}")

        def info(self, msg: object, *args: object, **kwargs: object) -> None:
            print(f"[INFO] {msg}")

        def warning(self, msg: object, *args: object, **kwargs: object) -> None:
            print(f"[WARN] {msg}")

        def error(self, msg: object, *args: object, **kwargs: object) -> None:
            print(f"[ERROR] {msg}")

    client = CamundaClient(logger=MyLogger())
    # endregion ReadmeCustomLogger


def readme_disable_logging() -> None:
    # region ReadmeDisableLogging
    from camunda_orchestration_sdk import CamundaClient, NullLogger

    client = CamundaClient(logger=NullLogger())
    # endregion ReadmeDisableLogging


# ---------- Heritable Worker Defaults ----------


async def readme_worker_defaults_env() -> None:
    from camunda_orchestration_sdk import CamundaAsyncClient, ConnectedJobContext, WorkerConfig

    async def handle_payment(job: ConnectedJobContext) -> dict[str, object]:
        return {"status": "paid"}

    async def handle_notification(job: ConnectedJobContext) -> dict[str, object]:
        return {"sent": True}

    async with CamundaAsyncClient() as client:
        # region ReadmeWorkerDefaultsEnv
        # No need to set job_timeout_milliseconds on every worker — inherited from env
        client.create_job_worker(
            config=WorkerConfig(job_type="payment-service"),
            callback=handle_payment,
        )
        client.create_job_worker(
            config=WorkerConfig(job_type="notification-service"),
            callback=handle_notification,
        )
        # endregion ReadmeWorkerDefaultsEnv


async def readme_worker_defaults_client() -> None:
    from camunda_orchestration_sdk import CamundaAsyncClient, ConnectedJobContext, WorkerConfig

    async def handle_payment(job: ConnectedJobContext) -> dict[str, object]:
        return {"status": "paid"}

    async def handle_shipping(job: ConnectedJobContext) -> dict[str, object]:
        return {"shipped": True}

    # region ReadmeWorkerDefaultsClient
    client = CamundaAsyncClient(configuration={
        "CAMUNDA_WORKER_TIMEOUT": "30000",
        "CAMUNDA_WORKER_MAX_CONCURRENT_JOBS": "16",
        "CAMUNDA_WORKER_NAME": "my-app",
    })

    # Both workers inherit timeout, concurrency, and name
    client.create_job_worker(
        config=WorkerConfig(job_type="payment-service"),
        callback=handle_payment,
    )
    client.create_job_worker(
        config=WorkerConfig(job_type="shipping-service"),
        callback=handle_shipping,
    )
    # endregion ReadmeWorkerDefaultsClient


# ---------- Job Corrections ----------


async def readme_job_corrections() -> None:
    # region ReadmeJobCorrections
    from camunda_orchestration_sdk import ConnectedJobContext
    from camunda_orchestration_sdk.models import (
        JobCompletionRequest,
        JobResultUserTask,
        JobResultCorrections,
    )

    async def validate_task(job: ConnectedJobContext) -> JobCompletionRequest:
        return JobCompletionRequest(
            result=JobResultUserTask(
                type_="userTask",
                corrections=JobResultCorrections(
                    assignee="corrected-user",
                    priority=80,
                ),
            ),
        )
    # endregion ReadmeJobCorrections


async def readme_job_corrections_denied() -> None:
    from camunda_orchestration_sdk import ConnectedJobContext
    from camunda_orchestration_sdk.models import (
        JobCompletionRequest,
        JobResultUserTask,
    )

    # region ReadmeJobCorrectionsDenied
    async def review_task(job: ConnectedJobContext) -> JobCompletionRequest:
        return JobCompletionRequest(
            result=JobResultUserTask(
                type_="userTask",
                denied=True,
                denied_reason="Insufficient documentation",
            ),
        )
    # endregion ReadmeJobCorrectionsDenied


# ---------- Job Logger ----------


async def readme_job_logger() -> None:
    from camunda_orchestration_sdk import ConnectedJobContext

    # region ReadmeJobLogger
    async def handler(job: ConnectedJobContext) -> dict[str, object]:
        job.log.info(f"Starting work on {job.job_key}")
        # ... do work ...
        job.log.debug("Work completed successfully")
        return {"done": True}
    # endregion ReadmeJobLogger


# ---------- Execution Strategies ----------


async def readme_execution_strategies() -> None:
    from camunda_orchestration_sdk import CamundaAsyncClient, WorkerConfig

    async with CamundaAsyncClient() as client:
        # region ReadmeExecutionStrategies
        from camunda_orchestration_sdk import SyncJobContext, JobContext

        # Force thread pool for a sync handler (receives SyncJobContext)
        def io_handler(job: SyncJobContext) -> dict[str, object]:
            return {"done": True}

        client.create_job_worker(
            config=WorkerConfig(job_type="io-bound-task", job_timeout_milliseconds=30_000),
            callback=io_handler,
            execution_strategy="thread",
        )

        # Force process pool for CPU-heavy work (receives plain JobContext)
        def cpu_handler(job: JobContext) -> dict[str, object]:
            return {"computed": True}

        client.create_job_worker(
            config=WorkerConfig(job_type="image-processing", job_timeout_milliseconds=120_000),
            callback=cpu_handler,
            execution_strategy="process",
        )
        # endregion ReadmeExecutionStrategies
